File: rust-wgpu-winit/lib/src/texture.rs
use std::sync::Arc;

use color_eyre::eyre::Result;
use glam::UVec2;
use image::DynamicImage;
use wgpu::{
    AddressMode, BindGroup, BindGroupDescriptor, BindGroupEntry, BindGroupLayout,
    BindGroupLayoutDescriptor, BindGroupLayoutEntry, BindingResource, BindingType, Device,
    Extent3d, FilterMode, Origin3d, Queue, SamplerBindingType, SamplerDescriptor, ShaderStages,
    TexelCopyBufferLayout, TexelCopyTextureInfo, TextureAspect, TextureDescriptor,
    TextureDimension, TextureFormat, TextureSampleType, TextureUsages, TextureViewDescriptor,
    TextureViewDimension,
};

#[derive(Debug, Clone, Copy)]
pub struct TextureBindings {
    pub texture: u32,
    pub sampler: u32,
}

pub struct Texture {
    device: Arc<Device>,
    queue: Arc<Queue>,
    binding: TextureBindings,
    size: Extent3d,
    texture: wgpu::Texture,
    bind_group: BindGroup,
}

impl Texture {
    pub fn with_size(
        device: Arc<Device>,
        queue: Arc<Queue>,
        bindings: TextureBindings,
        width: u32,
        height: u32,
    ) -> Result<Self> {
        Self::with_init_callback(device, queue, bindings, width, height, |_, _, _| Ok(()))
    }

    pub fn from_image(
        device: Arc<Device>,
        queue: Arc<Queue>,
        bindings: TextureBindings,
        image: &DynamicImage,
    ) -> Result<Self> {
        let image_rgba8 = match image.as_rgba8() {
            Some(x) => x,
            None => &image.to_rgba8(),
        };
        let width = image.width();
        let height = image.height();
        Self::with_init_callback(
            device,
            queue,
            bindings,
            width,
            height,
            |queue, texture, size| {
                queue.write_texture(
                    TexelCopyTextureInfo {
                        texture,
                        mip_level: 0,
                        origin: Origin3d::ZERO,
                        aspect: TextureAspect::All,
                    },
                    image_rgba8,
                    TexelCopyBufferLayout {
                        offset: 0,
                        bytes_per_row: Some(4 * width),
                        rows_per_image: Some(height),
                    },
                    size,
                );
                Ok(())
            },
        )
    }

    pub fn size(&self) -> &Extent3d {
        &self.size
    }

    pub fn width(&self) -> u32 {
        self.size.width
    }

    pub fn height(&self) -> u32 {
        self.size.height
    }

    pub fn bind_group(&self) -> &BindGroup {
        &self.bind_group
    }

    pub fn enqueue_update(&mut self, image: &DynamicImage, origin: UVec2) {
        let image_rgba8 = match image.as_rgba8() {
            Some(x) => x,
            None => &image.to_rgba8(),
        };
        let image_width = image.width();
        let image_height = image.height();
        self.queue.write_texture(
            TexelCopyTextureInfo {
                texture: &self.texture,
                mip_level: 0,
                origin: Origin3d {
                    x: origin.x,
                    y: origin.y,
                    z: 0,
                },
                aspect: TextureAspect::All,
            },
            image_rgba8,
            TexelCopyBufferLayout {
                offset: 0,
                bytes_per_row: Some(4 * image_width),
                rows_per_image: Some(image_height),
            },
            Extent3d {
                width: image_width,
                height: image_height,
                depth_or_array_layers: 1,
            },
        );
        self.bind_group = Self::create_bind_group_layout(
            &self.device,
            self.binding.texture,
            self.binding.sampler,
            &self.texture,
        );
    }

    fn with_init_callback(
        device: Arc<Device>,
        queue: Arc<Queue>,
        bindings: TextureBindings,
        width: u32,
        height: u32,
        f: impl FnOnce(&Queue, &wgpu::Texture, Extent3d) -> Result<()>,
    ) -> Result<Self> {
        let size = Extent3d {
            width,
            height,
            depth_or_array_layers: 1,
        };
        let texture = device.create_texture(&TextureDescriptor {
            label: None,
            size,
            mip_level_count: 1,
            sample_count: 1,
            dimension: TextureDimension::D2,
            format: TextureFormat::Rgba8UnormSrgb,
            usage: TextureUsages::TEXTURE_BINDING | TextureUsages::COPY_DST,
            view_formats: &[],
        });
        f(&queue, &texture, size)?;
        let bind_group =
            Self::create_bind_group_layout(&device, bindings.texture, bindings.sampler, &texture);
        Ok(Self {
            device,
            queue,
            binding: bindings,
            size,
            texture,
            bind_group,
        })
    }

    fn create_bind_group_layout(
        device: &Device,
        texture_binding: u32,
        sampler_binding: u32,
        texture: &wgpu::Texture,
    ) -> BindGroup {
        device.create_bind_group(&BindGroupDescriptor {
            label: None,
            layout: &bind_group_layout(device, texture_binding, sampler_binding),
            entries: &[
                BindGroupEntry {
                    binding: texture_binding,
                    resource: BindingResource::TextureView(
                        &texture.create_view(&TextureViewDescriptor::default()),
                    ),
                },
                BindGroupEntry {
                    binding: sampler_binding,
                    resource: BindingResource::Sampler(&device.create_sampler(
                        &SamplerDescriptor {
                            address_mode_u: AddressMode::ClampToEdge,
                            address_mode_v: AddressMode::ClampToEdge,
                            address_mode_w: AddressMode::ClampToEdge,
                            mag_filter: FilterMode::Linear,
                            min_filter: FilterMode::Linear,
                            mipmap_filter: FilterMode::Nearest,
                            ..Default::default()
                        },
                    )),
                },
            ],
        })
    }
}

pub fn bind_group_layout(
    device: &Device,
    texture_binding: u32,
    sampler_binding: u32,
) -> BindGroupLayout {
    device.create_bind_group_layout(&BindGroupLayoutDescriptor {
        label: None,
        entries: &[
            BindGroupLayoutEntry {
                binding: texture_binding,
                visibility: ShaderStages::FRAGMENT,
                ty: BindingType::Texture {
                    multisampled: false,
                    view_dimension: TextureViewDimension::D2,
                    sample_type: TextureSampleType::Float { filterable: true },
                },
                count: None,
            },
            BindGroupLayoutEntry {
                binding: sampler_binding,
                visibility: ShaderStages::FRAGMENT,
                ty: BindingType::Sampler(SamplerBindingType::Filtering),
                count: None,
            },
        ],
    })
}

File: csharp-silk-webgpu/Experiment/WebGPU/Texture.cs
namespace Experiment.WebGPU;

using System.Dynamic;
using System.Reflection;
using Silk.NET.Maths;
using Silk.NET.WebGPU;

public unsafe class Texture : IDisposable
{
	public class Description : IDisposable
	{
		public required VideoDriver VideoDriver { get; init; }
		public required BindGroupLayout* BindGroupLayout { get; init; }
		public required bool ReleaseBindGroupLayout { get; init; }
		public required uint TextureBinding { get; init; }
		public required uint SamplerBinding { get; init; }

		public void Dispose()
		{
			if (ReleaseBindGroupLayout)
			{
				VideoDriver.WebGPU.BindGroupLayoutRelease(BindGroupLayout);
			}
		}
	}

	public interface IDescriptionSource
	{
		public Description TextureDescription { get; }
	}

	private readonly VideoDriver videoDriver;
	private readonly Vector2D<int> size;
	private readonly Silk.NET.WebGPU.Texture* texture;
	private readonly BindGroup* bindGroup;

	public Texture(Description description, Vector2D<int> size)
	{
		try
		{
			this.videoDriver = description.VideoDriver;
			this.size = size;

			var textureDescriptor = new TextureDescriptor()
			{
				Size = new()
				{
					Width = (uint)size.X,
					Height = (uint)size.Y,
					DepthOrArrayLayers = 1,
				},
				MipLevelCount = 1,
				SampleCount = 1,
				Dimension = TextureDimension.Dimension2D,
				Format = TextureFormat.Rgba8Unorm,
				Usage = TextureUsage.TextureBinding | TextureUsage.CopyDst,
			};
			texture = videoDriver.WebGPU.DeviceCreateTexture(videoDriver.Device, ref textureDescriptor);

			var textureViewDescriptor = new TextureViewDescriptor()
			{
				Dimension = TextureViewDimension.Dimension2D,
				BaseMipLevel = 0,
				MipLevelCount = 1,
				BaseArrayLayer = 0,
				ArrayLayerCount = 1,
				Aspect = TextureAspect.All,
			};
			var textureView = videoDriver.WebGPU.TextureCreateView(texture, ref textureViewDescriptor);

			var samplerDescriptor = new SamplerDescriptor()
			{
				AddressModeU = AddressMode.ClampToEdge,
				AddressModeV = AddressMode.ClampToEdge,
				AddressModeW = AddressMode.ClampToEdge,
				MagFilter = FilterMode.Linear,
				MinFilter = FilterMode.Linear,
				MipmapFilter = MipmapFilterMode.Nearest,
				LodMinClamp = 1,
				LodMaxClamp = 1,
				Compare = CompareFunction.Undefined,
				MaxAnisotropy = 1,
			};
			var sampler = videoDriver.WebGPU.DeviceCreateSampler(videoDriver.Device, ref samplerDescriptor);

			bindGroup = videoDriver.CreateBindGroup(
				description.BindGroupLayout,
				[
					new()
					{
						Binding = description.TextureBinding,
						TextureView = textureView,
					},
					new()
					{
						Binding = description.SamplerBinding,
						Sampler = sampler,
					},
				]
			);

			videoDriver.WebGPU.TextureViewRelease(textureView);
			videoDriver.WebGPU.SamplerRelease(sampler);
		}
		finally
		{
			description.Dispose();
		}
	}

	internal BindGroup* BindGroup => bindGroup;

	public void Dispose()
	{
		videoDriver.WebGPU.TextureRelease(texture);
		videoDriver.WebGPU.BindGroupRelease(bindGroup);
	}

	public Vector2D<int> Size => size;

	public void QueueUpdate(SixLabors.ImageSharp.Image image, Vector2D<int> origin)
	{
		var imageConfiguration = image.Configuration.Clone();
		imageConfiguration.PreferContiguousImageBuffers = true;
		var imageInTheRightFormat = image.CloneAs<SixLabors.ImageSharp.PixelFormats.Rgba32>(imageConfiguration);
		if (!imageInTheRightFormat.DangerousTryGetSinglePixelMemory(out var memory))
		{
			throw new Exception("failed to get contiguous memory block for image");
		}
		using var memoryHandle = memory.Pin();
		var imageCopyTexture = new ImageCopyTexture()
		{
			Texture = texture,
			MipLevel = 0,
			Origin = new()
			{
				X = (uint)origin.X,
				Y = (uint)origin.Y,
				Z = 0,
			},
			Aspect = TextureAspect.All,
		};
		var rowLengthInBytes = imageInTheRightFormat.Width * 4;
		var dataLayout = new TextureDataLayout()
		{
			Offset = 0,
			BytesPerRow = (uint)rowLengthInBytes,
			RowsPerImage = (uint)imageInTheRightFormat.Height,
		};
		var writeSize = new Extent3D()
		{
			Width = (uint)imageInTheRightFormat.Width,
			Height = (uint)imageInTheRightFormat.Height,
			DepthOrArrayLayers = 1,
		};
		videoDriver.WebGPU.QueueWriteTexture(
			videoDriver.Queue,
			ref imageCopyTexture,
			memoryHandle.Pointer,
			(nuint)(imageInTheRightFormat.Height * rowLengthInBytes),
			ref dataLayout,
			ref writeSize
		);
	}
}

public static class TextureExtensions
{
	public static Texture CreateTexture(this Texture.IDescriptionSource source, Vector2D<int> size)
	{
		return new Texture(source.TextureDescription, size);
	}

	public static Texture CreateTexture(this Texture.IDescriptionSource source, SixLabors.ImageSharp.Image image)
	{
		var texture = source.CreateTexture(new Vector2D<int>(image.Width, image.Height));
		texture.QueueUpdate(image, new(0, 0));
		return texture;
	}

	public static Texture CreateTexture(this Texture.IDescriptionSource source, Stream stream)
	{
		using var image = SixLabors.ImageSharp.Image.Load(stream)
			?? throw new NullReferenceException("failed to load image from stream");
		return source.CreateTexture(image);
	}

	public static Texture CreateTextureFromManifestResource(this Texture.IDescriptionSource source, Assembly assembly, string name)
	{
		using var stream = assembly.AssertManifestResourceStream(name);
		return source.CreateTexture(stream);
	}

	public static Texture CreateTextureFromManifestResource(this Texture.IDescriptionSource source, string name)
	{
		return source.CreateTextureFromManifestResource(Assembly.GetExecutingAssembly(), name);
	}
}
